fearless/Common/Helpers/ChainAssetsFetching.swift

```diff
@@ -0,0 +1,365 @@
+        case assetName(SortOrder)
+        case chainName(SortOrder)
+        case isTest(SortOrder)
+        case isPolkadotOrKusama(SortOrder)
+        case assetId(SortOrder)
+    }
+
+    enum SortOrder {
+        case ascending
+        case descending
+    }
+
+    private enum Sort {
+        case price(SortOrder)
+        case balance(SortOrder, [ChainAsset: AccountInfo?])
+        case usdBalance(SortOrder, [ChainAsset: AccountInfo?])
+        case assetName(SortOrder)
+        case chainName(SortOrder)
+        case isTest(SortOrder)
+        case isPolkadotOrKusama(SortOrder)
+        case assetId(SortOrder)
+    }
+
+    private let chainRepository: AnyDataProviderRepository<ChainModel>
+    private let operationQueue: OperationQueue
+    private let meta: MetaAccountModel
+    private let accountInfoFetching: AccountInfoFetching
+
+    init(
+        chainRepository: AnyDataProviderRepository<ChainModel>,
+        accountInfoFetching: AccountInfoFetching,
+        operationQueue: OperationQueue = OperationQueue(),
+        meta: MetaAccountModel
+    ) {
+        self.chainRepository = chainRepository
+        self.operationQueue = operationQueue
+        self.accountInfoFetching = accountInfoFetching
+        self.meta = meta
+    }
+
+    func fetch(
+        filters: [Filter],
+        sortDescriptors: [SortDescriptor],
+        completionBlock: @escaping (Result<[ChainAsset], Error>?) -> Void
+    ) {
+        let operation = chainRepository.fetchAllOperation(with: .none)
+        operation.completionBlock = { [weak self] in
+            guard let strongSelf = self else {
+                return
+            }
+            switch operation.result {
+            case let .success(chains):
+                var chainAssets: [ChainAsset] = []
+                chains.forEach { chain in
+                    chainAssets.append(contentsOf: chain.chainAssets)
+                }
+
+                chainAssets = strongSelf.filter(chainAssets: chainAssets, filters: filters)
+
+                if sortDescriptors.contains(where: { sort in
+                    switch sort {
+                    case .balance, .usdBalance:
+                        return true
+                    default:
+                        return false
+                    }
+                }) {
+                    var accountInfos: [ChainAsset: AccountInfo?] = [:]
+                    chainAssets.forEach { [weak self] chainAsset in
+                        guard let strongSelf = self,
+                              let accountId = strongSelf.meta.fetch(for: chainAsset.chain.accountRequest())?.accountId else {
+                            return
+                        }
+
+                        strongSelf.accountInfoFetching.fetch(
+                            for: chainAsset,
+                            accountId: accountId
+                        ) { chainAsset, accountInfo in
+                            accountInfos[chainAsset] = accountInfo
```

Comment:
there may be a delay and strongSelf.convertSorts will be run without accountInfo